### PR TITLE
Includes a minimal authentication step before SFU and Etherpad connections

### PR DESCRIPTION
**Pull request by pedrobmarin**
Resolved #6506 conflicts and included validation to Etherpad connection

For this to work we need to change `notes.nginx` for:
```
# https://github.com/ether/etherpad-lite/wiki/How-to-put-Etherpad-Lite-behind-a-reverse-Proxy
location ~ "^\/pad\/p\/(\w+)" {
    rewrite                /pad/(.*) /$1 break;
    rewrite                ^/pad$ /pad/ permanent;
    proxy_pass             http://localhost:9001;
    proxy_pass_header      Server;
    proxy_redirect         / /pad;
    proxy_set_header       Host $host;
    proxy_buffering off;

    auth_request /bigbluebutton/connection/checkAuthorization;
    auth_request_set $auth_status $upstream_status;
}

location /pad {
    rewrite                /pad/(.*) /$1 break;
    rewrite                ^/pad$ /pad/ permanent;
    proxy_pass             http://localhost:9001/;
    proxy_pass_header      Server;
    proxy_redirect         / /pad;
    proxy_set_header       Host $host;
    proxy_buffering off;
}

location /pad/socket.io {
    rewrite /pad/socket.io/(.*) /socket.io/$1 break;
    proxy_pass http://localhost:9001/;
    proxy_redirect         / /pad/;
    proxy_set_header Host $host;
    proxy_buffering off;
    proxy_set_header X-Real-IP $remote_addr;  # http://wiki.nginx.org/HttpProxyModule
    proxy_set_header X-Forwarded-For $remote_addr; # EP logs to show the actual remote IP
    proxy_set_header X-Forwarded-Proto $scheme; # for EP to set secure cookie flag when https is used
    proxy_set_header Host $host;  # pass the host header
    proxy_http_version 1.1;  # recommended with keepalive connections
    # WebSocket proxying - from http://nginx.org/en/docs/http/websocket.html
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
}

location /static {
    rewrite /static/(.*) /static/$1 break;
    proxy_pass http://localhost:9001/;
    proxy_set_header Host $host;
    proxy_buffering off;
}
```
The first block is the big difference. It will auth_request https://HOSTNAME/pad/p/NOTE_ID

and `webrtc-sfu.nginx` should look like:
```
location /bbb-webrtc-sfu {
        proxy_pass http://127.0.0.1:3008;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_read_timeout 6h;
        proxy_send_timeout 6h;
        client_body_timeout 6h;
        send_timeout 6h;

        auth_request /bigbluebutton/connection/checkAuthorization;
        auth_request_set $auth_status $upstream_status;
}
```

This modifications should unauthorize any connections to this locations that append an invalid sessionToken. Etherpad can be tested by trying to access it outside BBB client. I tested SFU connections by modifying https://github.com/pedrobmarin/bigbluebutton/blob/validate-sfu-conn/bigbluebutton-html5/imports/ui/services/auth/index.js#L246 to send an invalid sessionToken. This should output a 401 exception in console.

And, of course, `web-nginx` should include
```
		location = /bigbluebutton/connection/checkAuthorization {
			internal;
			proxy_pass               http://127.0.0.1:8080;
			proxy_pass_request_body  off;
			proxy_set_header         Content-Length "";
			proxy_set_header         X-Original-URI $request_uri;
		}
```

### CODE REVIEW

File: bigbluebutton-web/grails-app/controllers/org/bigbluebutton/web/controllers/ConnectionController.groovy [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import org.bigbluebutton.api.domain.UserSession
+import org.bigbluebutton.api.util.ParamsUtil
+
+class ConnectionController {
+  MeetingService meetingService
+
+  def checkAuthorization = {
+    try {
+      def uri = request.getHeader("x-original-uri")
+      def sessionToken = ParamsUtil.getSessionToken(uri)
```
Comment (capilkey) [inline bigbluebutton-web/grails-app/controllers/org/bigbluebutton/web/controllers/ConnectionController.groovy]: Can you not just get sessionToken directly from `params` like the other controllers? https://github.com/bigbluebutton/bigbluebutton/blob/460c81686819a01e35d35e5eb1594ea87fc08c1c/bigbluebutton-web/grails-app/controllers/org/bigbluebutton/web/controllers/ApiController.groovy#L1416

Comment (pedrobmarin) [reply bigbluebutton-web/grails-app/controllers/org/bigbluebutton/web/controllers/ConnectionController.groovy]: I tried but did not succeeded when it comes in a wss connection, the SFU case.

Comment (capilkey) [reply bigbluebutton-web/grails-app/controllers/org/bigbluebutton/web/controllers/ConnectionController.groovy]: Strange, but if you tried it then I guess something's different.

Comment (pedrobmarin) [reply bigbluebutton-web/grails-app/controllers/org/bigbluebutton/web/controllers/ConnectionController.groovy]: Well, I'm no nginx expert... Maybe it is possible but I missed on something else.

**Comment by fcecagno** [comment]
Just copying from #6506, this is what needs to be added to the nginx config files of etherpad and bbb-webrtc-sfu in order to enable authentication to the component:

```
  auth_request /bigbluebutton/connection/checkAuthorization;
  auth_request_set $auth_status $upstream_status;
```

**Comment by antobinary** [comment]
Do we need to change https://github.com/bigbluebutton/bigbluebutton/blob/master/bigbluebutton-html5/client/compatibility/kurento-extension.js#L46 ?

**Comment by pedrobmarin** [comment]
> Do we need to change https://github.com/bigbluebutton/bigbluebutton/blob/master/bigbluebutton-html5/client/compatibility/kurento-extension.js#L46 ?

@antobinary I would say no, but I'm not completely sure right now. I'm revising this pull request. Checking if I can make some changes on code quality like @capilkey suggested. I'll ping you when I think this is good enough.

**Comment by pedrobmarin** [comment]
Ok, tried this morning to change the way I get the session token with no success. If anyone has a sugestion, I'm all ears.

About testing, AFAIK this pull request includes all the client modification needed. And that's adding a sessionToken argument in SFU and Etherpad connections. Nginx should do all the heavy lifting from here. For this to work we need to change `notes.nginx` for:
```
# https://github.com/ether/etherpad-lite/wiki/How-to-put-Etherpad-Lite-behind-a-reverse-Proxy
location ~ "^\/pad\/p\/(\w+)" {
    rewrite                /pad/(.*) /$1 break;
    rewrite                ^/pad$ /pad/ permanent;
    proxy_pass             http://localhost:9001;
    proxy_pass_header      Server;
    proxy_redirect         / /pad;
    proxy_set_header       Host $host;
    proxy_buffering off;

    auth_request /bigbluebutton/connection/checkAuthorization;
    auth_request_set $auth_status $upstream_status;
}

location /pad {
    rewrite                /pad/(.*) /$1 break;
    rewrite                ^/pad$ /pad/ permanent;
    proxy_pass             http://localhost:9001/;
    proxy_pass_header      Server;
    proxy_redirect         / /pad;
    proxy_set_header       Host $host;
    proxy_buffering off;
}

location /pad/socket.io {
    rewrite /pad/socket.io/(.*) /socket.io/$1 break;
    proxy_pass http://localhost:9001/;
    proxy_redirect         / /pad/;
    proxy_set_header Host $host;
    proxy_buffering off;
    proxy_set_header X-Real-IP $remote_addr;  # http://wiki.nginx.org/HttpProxyModule
    proxy_set_header X-Forwarded-For $remote_addr; # EP logs to show the actual remote IP
    proxy_set_header X-Forwarded-Proto $scheme; # for EP to set secure cookie flag when https is used
    proxy_set_header Host $host;  # pass the host header
    proxy_http_version 1.1;  # recommended with keepalive connections
    # WebSocket proxying - from http://nginx.org/en/docs/http/websocket.html
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
}

location /static {
    rewrite /static/(.*) /static/$1 break;
    proxy_pass http://localhost:9001/;
    proxy_set_header Host $host;
    proxy_buffering off;
}
```
The first block is the big difference. It will auth_request https://HOSTNAME/pad/p/NOTE_ID

and `webrtc-sfu.nginx` should look like:
```
location /bbb-webrtc-sfu {
        proxy_pass http://127.0.0.1:3008;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_read_timeout 6h;
        proxy_send_timeout 6h;
        client_body_timeout 6h;
        send_timeout 6h;

        auth_request /bigbluebutton/connection/checkAuthorization;
        auth_request_set $auth_status $upstream_status;
}
```

This modifications should unauthorize any connections to this locations that append an invalid sessionToken. Etherpad can be tested by trying to access it outside BBB client. I tested SFU connections by modifying https://github.com/pedrobmarin/bigbluebutton/blob/validate-sfu-conn/bigbluebutton-html5/imports/ui/services/auth/index.js#L246 to send an invalid sessionToken. This should output a 401 exception in console.

File: bigbluebutton-html5/imports/ui/components/video-provider/component.jsx [hunk 2]
```diff
@@ -110,7 +111,7 @@ class VideoProvider extends Component {
     };
 
     // Set a valid bbb-webrtc-sfu application server socket in the settings
-    this.ws = new ReconnectingWebSocket(Meteor.settings.public.kurento.wsUrl);
+    this.ws = new ReconnectingWebSocket(Auth.authenticateURL(Meteor.settings.public.kurento.wsUrl));
```
Comment (antobinary) [inline bigbluebutton-html5/imports/ui/components/video-provider/component.jsx]: If I revert this line I am still able to use webcam :man_shrugging: Doesn't this mean we didn't auth?
I have applied the nginx and tomcat changes.

Comment (antobinary) [reply bigbluebutton-html5/imports/ui/components/video-provider/component.jsx]: I was missing 2 lines in the sfu nginx config

File: bbb-common-web/src/main/java/org/bigbluebutton/api/util/ParamsUtil.java [hunk 3]
```diff
@@ -21,4 +28,23 @@ public static boolean isValidMeetingId(String meetingId) {
   public static boolean containsChar(String text, String chars) {
     return StringUtils.containsAny(text, chars);
   }
+
+  public static String getSessionToken(String url) {
+    String token = "undefined";
+    try {
+      String decodedURL = URLDecoder.decode(url, "UTF-8");
+      String[] splitedURL = decodedURL.split("\\?");
```
Comment (antobinary) [inline bbb-common-web/src/main/java/org/bigbluebutton/api/util/ParamsUtil.java]: Consider renaming `splitedURL` to `splitURL`

File: bigbluebutton-web/grails-app/controllers/org/bigbluebutton/web/controllers/ConnectionController.groovy (hunk 1 above)
Comment (ritzalam) [reply bigbluebutton-web/grails-app/controllers/org/bigbluebutton/web/controllers/ConnectionController.groovy]: I think this should work for now. We can change this later if we figure out a better way.

**Pull request closed without merging**